prover/src/lib.rs

```diff
@@ -192,44 +194,73 @@ pub trait Prover {
     /// secret and public inputs. Public inputs must match the value returned from
     /// [Self::get_pub_inputs()](Prover::get_pub_inputs) for the provided trace.
     fn prove(&self, trace: Self::Trace) -> Result<StarkProof, ProverError> {
-        // figure out which version of the generic proof generation procedure to run. this is a sort
-        // of static dispatch for selecting two generic parameter: extension field and hash
-        // function.
-        match self.options().field_extension() {
-            FieldExtension::None => self.generate_proof::<Self::BaseField>(trace),
-            FieldExtension::Quadratic => {
-                if !<QuadExtension<Self::BaseField>>::is_supported() {
-                    return Err(ProverError::UnsupportedFieldExtension(2));
-                }
-                self.generate_proof::<QuadExtension<Self::BaseField>>(trace)
-            }
-            FieldExtension::Cubic => {
-                if !<CubeExtension<Self::BaseField>>::is_supported() {
-                    return Err(ProverError::UnsupportedFieldExtension(3));
-                }
-                self.generate_proof::<CubeExtension<Self::BaseField>>(trace)
-            }
-        }
-    }
+        // FIXME: the `self.options().field_extension` is now irrelevant and could be inconsistent
+        // with `E`.
+        assert!(!trace.info().is_multi_segment());
 
-    // HELPER METHODS
-    // --------------------------------------------------------------------------------------------
+        // 0 ----- instantiate AIR and prover channel ---------------------------------------------
 
-    /// Performs the actual proof generation procedure, generating the proof that the provided
-    /// execution `trace` is valid against this prover's AIR.
-    /// TODO: make this function un-callable externally?
-    #[doc(hidden)]
-    fn generate_proof<E>(&self, mut trace: Self::Trace) -> Result<StarkProof, ProverError>
+        // serialize public inputs; these will be included in the seed for the public coin
+        let pub_inputs = self.get_pub_inputs(&trace);
+        let pub_inputs_elements = pub_inputs.to_elements();
+
+        // create an instance of AIR for the provided parameters. this takes a generic description
+        // of the computation (provided via AIR type), and creates a description of a specific
+        // execution of the computation for the provided public inputs.
+        let air = Self::Air::new(trace.info().clone(), pub_inputs, self.options().clone());
+
+        // create a channel which is used to simulate interaction between the prover and the
+        // verifier; the channel will be used to commit to values and to draw randomness that
+        // should come from the verifier.
+        let mut channel =
+            ProverChannel::<Self::Air, Self::BaseField, Self::HashFn, Self::RandomCoin>::new(
+                &air,
+                pub_inputs_elements,
+            );
```

Comment:
We can't pull this out into a new function, since `channel` borrows `air`, and Rust won't let us return a value that is borrowed